### ISSUE TITLE
Investigate ExUnit performance

### ISSUE THREAD

**Issue by bjfish** [elixir-lang/elixir#4461]
I was investigating `mix test` run for the Elixir's ExUnit tests with `:observer.start`.

I found the following utilization during the test.
![screen shot 2016-03-31 at 11 29 09 am](https://cloud.githubusercontent.com/assets/110023/14186555/91c07ca4-f744-11e5-8de4-6295a02daa29.png)
The drop off to almost zero utilization in the middle of the graph is after `ParallelRequire` finishes requiring all test files I believe. So, my initial performance idea is that tests could start to run after each test file finishes requiring using the `ParallelRequire` callback. `ParallelRequire` happens here: https://github.com/elixir-lang/elixir/blob/master/lib/mix/lib/mix/tasks/test.ex

I also found MsgQ with 1-3 messages but I'm told this is normal.

For reference, the tests report a run time for me of `Finished in 28.0 seconds (15.0s on load, 12.9s on tests)` so it spent more time on load then in tests.

What other options are there for improving test performance?

### Environment

* Elixir version (elixir -v): Elixir 1.3.0-dev (dc8a804)
* Operating system: Mac OS X

**Comment by bjfish** [comment]
I did a proof of concept where files are required in parallel with running the tests. On its own, this change did not significantly improve performance or utilization.  It is nice to see the test suite starting immediately with no delay in seeing the first test success `.`.

I did find test performance/utilization being affected by the number of cases that could be run at a given time (`max cases` option). So, my next performance hypothesis will be to try to control test load via a pool of workers to make `max cases` dynamic at a given moment.  I'm not familiar with backpressure/load control so any tips would be appreciated as I try to build this.

**Comment by bjfish** [comment]
I've rewritten my previous proof of concept (POC) of requiring tests in parallel while running tests and have now found an estimated improved run times by  about 10% when tuned with `max_cases 32` and  about 15% when adding `+A 64`.

I think I'm close to concluding this investigation. There's just a two more questions I'd like to answer first:

- Is there an efficient way to get `@ex_unit_async` or `@moduletag async: async` value from a test module after compilation? Or is there change I could make to get this efficiently? I just need to know if a given module is a test or not and if so if it is async or sync. Right now, I'm calling the ExUnit.Server to get this information.

This is the patch for this POC: https://gist.github.com/bjfish/c1f08e7d96f9ee5128101a36810adb27

I'm also curious what the second IO spike is here:
![screen shot 2016-04-07 at 9 15 31 am](https://cloud.githubusercontent.com/assets/110023/14356181/6f842c0c-fca9-11e5-9bc4-3e39aa769150.png)

**Comment by josevalim** [comment]
Great job @bjfish! Are the benefits only for increasing max_cases or when you both increase max_cases and allow tests to start immediately?

Btw, currently there is no way to get all async cases.

**Comment by bjfish** [comment]
Changing max cases alone doesn't appear to have much of an effect. I measured the following timings on the Elixir test suite:
```
8 max cases 
mean {30.205, 32.958, 36.577, 35.826, 35.356} = 34.1844 seconds

16 max cases
mean {34.909, 34.444, 30.420, 33.683, 35.171} = 33.7254 seconds

32 max cases
mean {33.363, 33.882, 34.021, 34.763, 35.415} = 34.2888 seconds
```

**Comment by paulcsmith** [comment]
FWIW even if the tests finished at about the same time, getting the first `.` out faster would definitely make the perceived time go down, which would be a win in and of itself. So I think running as soon as the test is ready is an awesome idea!

Great work on this @bjfish

**Comment by bjfish** [comment]
I think it could be a good change to compile and test in parallel. However, I understand if this is not a priority at the moment. 

I'm not currently working on this so feel free to close this issue if you would like.

**Comment by josevalim** [comment]
> FWIW even if the tests finished at about the same time, getting the first . out faster would definitely make the perceived time go down, which would be a win in and of itself. So I think running as soon as the test is ready is an awesome idea!

This is a very good point. I can think of two downsides:

1. we will no longer be able to measure run time and load time (as they now happen together)

2. there is a chance you will run into concurrency issues. let's say you depend on a module defined in another test file. This is obviously not something you should do but you may start running into it anyway

Honestly, I don't think those two downsides matter much and I think we should likely proceed with this feature. What are your thoughts @bjfish and @paulcsmith?

**Comment by bjfish** [comment]
@josevalim 
I didn't run into problem 2 for when I tried this with elixir, eex, mix, ex_doc, etc.

I don't see much of a downside to number 1. I think the load time could just be counted as part of the test execution time; it's implied by being an exs file.

**Comment by paulcsmith** [comment]
I think number 1 shouldn't be an issue and I think number 2 is only an issue if there was something you could do about the load time, which I think besides deleting tests, there's not much that can be done.

So I think running tests as they're loaded would be awesome!

**Comment by josevalim** [comment]
Master will now print tests as they are loaded. Feedback is welcome!
